frame/system-token-manager/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+	) -> Option<SystemTokenId>;
+	/// Adjust the vote weight calculating exchange rate.
+	fn adjusted_weight(system_token: SystemTokenId, vote_weight: VoteWeight) -> VoteWeight;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::{OptionQuery, *};
+	use frame_system::{
+		pallet_prelude::*, SystemTokenId, SystemTokenMetadata, WrappedSystemTokenId,
+	};
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The string limit for name and symbol of system token.
+		#[pallet::constant]
+		type StringLimit: Get<u32>;
+		/// Max number which can be used as system tokens on parachain.
+		#[pallet::constant]
+		type MaxWrappedSystemToken: Get<u32>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An asset was on the relay chain.
+		AssetRegistered {
+			system_token_id: SystemTokenId,
+			system_token_metadata: SystemTokenMetadata,
+		},
+
+		AssetRemoved {
+			system_token_id: SystemTokenId,
+			system_token_metadata: SystemTokenMetadata,
+		},
+
+		AssetConverted {
+			wrapped_system_token: WrappedSystemTokenId,
+			system_token_id: SystemTokenId,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AssetAlreadyRegistered,
+		AssetNotRegistered,
+	}
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn relay_asset_link)]
+	/// List for original system token and metadata.
+	pub(super) type SystemTokenList<T: Config> =
+		StorageMap<_, Twox64Concat, SystemTokenId, SystemTokenMetadata, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn system_token_on_parachain)]
+	/// Coverter between WrappedSystemTokenId and original SystemTokenId.
+	pub(super) type SystemTokenOnParachain<T: Config> =
+		StorageMap<_, Twox64Concat, WrappedSystemTokenId, SystemTokenId, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn system_token_on_parachain_by_para_id)]
+	/// System token list for specific parachain by ParaId.
+	pub(super) type SystemTokenOnParachainByParaId<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		ParaId,
+		BoundedVec<WrappedSystemTokenId, T::MaxWrappedSystemToken>,
+		OptionQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn allowed_system_token)]
+	/// Wrapped System token list used in parachains.
+	pub(super) type AllowedSystemToken<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		ParaId,
+		Twox64Concat,
+		PalletId,
+		WrappedSystemTokenId,
+		OptionQuery,
+	>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register a system token
+		#[pallet::call_index(0)]
+		#[pallet::weight(1_000)]
+		pub fn register_system_token(
+			origin: OriginFor<T>,
+			system_token_id: SystemTokenId,
+			system_token_metadata: SystemTokenMetadata,
```

Comment:
Metadata 를 옵션으로 하는건 어떤가요?

Reply:
I think concrete data type is better for SysteomTokenMetadata. Because system tokens are specially important on infra block space, metadata should be provided before accepting as a system token.